community/kernel/src/main/java/org/neo4j/kernel/impl/store/kvstore/KeyValueStoreState.java

```diff
@@ -0,0 +1,222 @@
+/**
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.store.kvstore;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+class KeyValueStoreState<Key, Meta>
+{
+    private final AbstractKeyValueStore<Key, Meta, ?> logic;
+    private final KeyValueStoreFile<Meta> store;
+    private final ConcurrentMap<Key, byte[]> changes = new ConcurrentHashMap<>();
+    final File file;
+
+    KeyValueStoreState( AbstractKeyValueStore<Key, Meta, ?> logic, KeyValueStoreFile<Meta> store, File file )
+    {
+        this.logic = logic;
+        this.store = store;
+        this.file = file;
+    }
+
+    @SuppressWarnings("SynchronizationOnLocalVariableOrMethodParameter")
+    public void apply( final AbstractKeyValueStore.Update<Key> update ) throws IOException
+    {
+        byte[] value = changes.get( update.key );
+        if ( value == null )
+        {
+            final byte[] proposal = new byte[logic.valueSize];
+            synchronized ( proposal )
```

Comment:
Same as the sync on `value` below. To ensure that only one writer is changing this data at any given time. With the `proposal` being put into the map before writing to it, another thread could access it concurrently...

Putting it into the map before writing is how we ensure that we only have to invoke the write logic once.